sql/api/src/main/scala/org/apache/spark/sql/types/StringType.scala

```diff
@@ -106,3 +109,20 @@ case object StringType extends StringType(0) {
     new StringType(collationId)
   }
 }
+
+/**
+ * The result type of literals, column definitions without explicit collation, casts to string
+ * and some expressions that produce strings but whose output type is not based on the types of its
+ * children.
+ * Idea is to have this behave like a string with the default collation of the session, but that
+ * we can still differentiate it from a regular string type, because in some places default string
+ * is not the one with the session collation (e.g. in DDL commands).
+ */
+private[spark] class DefaultStringType extends StringType(0) {
```

Comment:
how is this different from `case object StringType extends StringType(0)`?

Reply:
case object StringType will have collationId = _collationId, where DefaultStringType will have SQLconf.get.defaultStringType.collationId

Reply:
ok, so this is kind of `StringType(0)` with a special marker to indicate that it should be replaced by a default collation later. Can we use a special collation id for this purpose?

Reply:
Exactly, it should be replaced in the analyzer  for DDL operations; for others it should just behave as the underlying session collation. I spent some time thinking  about it and haven't been able to figure out a better solution for this, but let me know if you think something else might work better.

I added the special id as requested.